README.md
# psil

Psil (Pronounced Sil) is a mini programming language to evaluate arithmetic expressions

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar psil-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/psil/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/psil/compare/0.1.0...0.1.1


test/psil/core_test.clj
(ns psil.core-test
  (:require [clojure.test :refer :all]
            [psil.core :refer :all]))

(deftest run-src-test
  (testing "Psil run-src"
    ;; -
    (is (= 0 (run-src "(- 2 (* 1 2))")))
    (is (= -2 (run-src "(- 2)")))
    (is (= 1 (run-src "(- 2 1)")))
    (is (= 1 (run-src "(- 2 (+ 1))")))
    (is (= 2 (run-src "(* 2 (- 3 1 1))")))
    (is (= 4 (run-src "(bind a (- 2)) (bind b (- 2)) (* a b)")))

    ;; /
    (is (= 1 (run-src "(/ 2 (* 1 2))")))
    (is (= 1/2 (run-src "(/ 2)")))
    (is (= 2 (run-src "(/ 2 1)")))
    (is (= 4 (run-src "(* 2 (/ 4 2))")))
    (is (= 3/2 (run-src "(bind a (/ 3 1 2))")))
    (is (= 3 (run-src "(* (bind a (/ 3 1 2)) 2)")))

    ;; single
    (is (= 123 (run-src "123")))
    (is (= 0 (run-src "(+)")))
    (is (= 1 (run-src "(*)")))
    (is (= 1 (run-src "(+ 1)")))
    (is (= 1 (run-src "(* 1)")))

    ;; misc
    (is (= 2 (run-src "(* 1 1 1 1 2)")))
    (is (= 36 (run-src "(* 2 (+ 1 2 3) 3)")))
    (is (= 6 (run-src "(* 2 (+ 1 ) 3)")))
    (is (= 192 (run-src "(* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))")))
    (is (= 192 (run-src "(* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))")))
    (is (= 192 (run-src "(bind a (* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2))))))")))
    (is (= 200 (run-src "(bind a (* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))) (+ 8 a)")))
    (is (= 10 (run-src "(bind a 1) (bind b 2) (+ 7 a b)")))
    (is (= 15 (run-src "(+ 1\n    (* 2 3)\n    (* 4 2))")))
    (is (= 12 (run-src "(bind radius 12)")))
    (is (= 100 (run-src "(bind length 10)\n(bind breadth 10)\n(* length breadth)")))

    ;; bind
    (is (= 100 (run-src "(bind length 10) (+ 1 2 3 4) (bind breadth 10)\n(* length breadth)")))
    (is (= 121 (run-src "(bind length 10)\n(bind breadth (+ length 1))\n(bind length 11)\n(* length breadth)")))
    (is (= 21 (run-src "(bind a 10)\n(bind b a)\n(bind a 11)\n(+ a b)")))

    ;; invalid
    (is (thrown? RuntimeException (run-src "(bind a 1) (bind b 2) (+ c a b)")))
    (is (thrown? RuntimeException (run-src "(bind a 1) (bind b 2) (+ c a b)")))
    (is (thrown? RuntimeException (run-src "(bind a a) (bind b 2) (+ c a b)")))
    (is (thrown? RuntimeException (run-src "()")))
    (is (thrown? RuntimeException (run-src "(1)")))
    (is (thrown? RuntimeException (run-src "a")))
    (is (thrown? RuntimeException (run-src "( a +")))
    (is (thrown? RuntimeException (run-src "( 1 + )")))
    (is (thrown? RuntimeException (run-src "(1 2 3 4 5)")))
    (is (thrown? RuntimeException (run-src "(+ 1) 2)")))
    (is (thrown? RuntimeException (run-src "(+ (* 1 2) + (* 3 4))")))

    ;; negative and fractional
    (is (= -1 (run-src "(- 1)")))
    (is (= 0 (run-src "(+ -1 1)")))
    (is (= 2.2 (run-src "(+ 1.2 1)")))
    (is (= 1.3 (run-src "(* 1.3 1)")))
    (is (= 1.2 (run-src "(bind a 1.2)")))
    (is (= 1 (run-src "(* -1 -1)")))
    (is (= -3.5999999999999996 (run-src "(* -3 1.2)")))
    (is (= -5.0 (run-src "(bind a -1.5)\n(bind b 2)\n(bind a -2.5)\n(* a b)")))

    ;; invalid
    (is (thrown? RuntimeException (run-src "(- 1..2)")))
    (is (thrown? RuntimeException (run-src "(- 2.4.)")))

    ))


src/psil/core.clj
(ns psil.core
  (:require [clojure.java.io :as io]
            [clojure.string :as s]
            [com.stuartsierra.component :as component]
            [psil.components :as comps])
  (:gen-class))

(defn bind-fn [variable-lookup-map args]
  (swap! variable-lookup-map assoc (first args) (second args))
  (@variable-lookup-map (first args)))


(defn +-fn [_ args]
  (apply + args))


(defn *-fn [_ args]
  (apply * args))


(defn --fn [_ args]
  (apply - args))


(defn div-fn [_ args]
  (apply / args))


(def symbol-map {"+" +-fn "*" *-fn "-" --fn "/" div-fn "bind" bind-fn})


(defn xsymbol? [s]
  (contains?
    (set (keys symbol-map)) s))


(defn to-number [s]
  (try (if (.contains s ".")
         (Double/parseDouble s)
         (Long/parseLong s))
       (catch NumberFormatException nfe false)))


(defn variable? [s]
  (->> s
       (every? #(or (and (>= (int %) 65) (<= (int %) 90))
                    (and (>= (int %) 97) (<= (int %) 122))))))

(defn raise []
  (throw (RuntimeException. "Invalid Program")))


(defn not-emp [stack]
  (not (.isEmpty stack)))


(defn strip-line-breaks-tabs [x]
  (s/replace x #"[\n\t\r]" " "))


(defn tokenize-s-expr-str [s-exp-str]
  (s/split (.trim s-exp-str) #"\s+"))


(defn tokenize-source [src]
  (filter (comp not empty?) (s/split (s/replace (s/replace src #"\(" " ( ") #"\)" " ) ") #"\s+")))


(defn parse-raw-source
  "Parses input source to produce tokenized code sequence
  (parse-file \"p7.psil\")
  => (\"(\" \"bind\" \"length\" \"10\" \")\" \"(\" \"bind\" \"breadth\" \"10\" \")\" \"(\" \"*\" \"length\" \"breadth\" \")\")\n
  "
  [raw-source]
  (->> raw-source
       strip-line-breaks-tabs
       tokenize-source))


(defn parse-source
  "Processes token sequence to segregate S-Expressions, also checks S-Exp
  syntacs and raise error if invalid. Produces a seq of S-Exp strings

  (parse-source (parse-file \"p7.psil\"))
  => [\"( bind length 10 )\" \"( bind breadth 10 )\" \"( * length breadth )\"]"
  [source]
  (let [stack (java.util.Stack.)
        sb (StringBuilder.)
        result (java.util.ArrayList.)
        pop-till-balanced (fn [s]
                            (if (.isEmpty s) (raise))
                            (while (and (not-emp s) (not (= (.peek s) "(")))
                              (.pop s))
                            (if (not-emp s) (.pop s) (raise)))]
    (doseq [token source]
      (cond
        (= token "(") (do (.append sb (str token " ")) (.push stack token))
        (or (xsymbol? token) (to-number token) (variable? token)) (.append sb (str token " "))
        (= token ")") (do (.append sb (str token " ")) (pop-till-balanced stack))
        :else (raise))
      (if (.isEmpty stack)
        (do (.add result (.trim (.toString sb)))
            (.setLength sb 0))))

    (if (not-emp stack)
      (if (not (to-number (.peek stack)))
        (raise)
        (.add result (.pop stack)))
      (vec result))))


(defn handle-number [stack number]
  (if (and (not-emp stack) (or (= (.peek stack) "(") (= (.peek stack) ")")))
    (raise)
    (if (.isEmpty stack)
      (raise)
      (.push stack (to-number number)))))


(defn handle-symb [stack symb]
  (if (.isEmpty stack)
    (raise)
    (if (not= (.peek stack) "(")
      (raise)
      (.push stack symb))))


(defn resolve-var [variable-lookup-map stack var]
  (if (= (.peek stack) "bind")
    var
    (if (contains? @variable-lookup-map var)
      (@variable-lookup-map var)
      (raise))))


(defn handle-var [variable-lookup-map stack var]
  (if (not-emp stack)
    (if (or (= (.peek stack) "(")
            (= (.peek stack) ")"))
      (raise)
      (.push stack (resolve-var variable-lookup-map stack var)))
    (raise)))


(defn handle-close-brack [variable-lookup-map stack]
  (let [operands (java.util.ArrayList.)]
    (if (.isEmpty stack) (raise))
    (while (and (not-emp stack) (not (= (.peek stack) "(")))
      (.add operands (.pop stack)))
    (if (not-emp stack)
      (do (.pop stack)
          (if (empty? operands) (raise))
          (let [ops-reversed (reverse operands)]
            (.push stack
                   ((symbol-map (first ops-reversed)) variable-lookup-map (rest ops-reversed)))))
      (raise))))


(defn eval-s-exp [variable-lookup-map s-exp-str]
  "Evaluates a single S-Exp string

  (eval-s-exp \"123\")
  => 123"
  (let [s (java.util.Stack.)
        tokens (tokenize-s-expr-str s-exp-str)
        cnt (count tokens)]
    (cond
      (< cnt 1) (raise)
      (= cnt 1) (if (to-number (first tokens))
                  (.push s (to-number (first tokens)))
                  (raise))
      :else (doseq [token tokens]
              (cond
                (= token "(") (.push s token)
                (xsymbol? token) (handle-symb s token)
                (to-number token) (handle-number s token)
                (variable? token) (handle-var variable-lookup-map s token)
                (= token ")") (handle-close-brack variable-lookup-map s)
                :else (raise))))
    (let [result (.pop s)]
      (if (not-emp s) (raise) result))))


(defn eval-program
  "Evaluates the sequence of S-Exps strings and returns the value of the last one"
  [variable-lookup-map s-exp-str-seq]
  (doseq [s-exp-str (drop-last s-exp-str-seq)]
    (eval-s-exp variable-lookup-map s-exp-str))
  (eval-s-exp variable-lookup-map (last s-exp-str-seq)))


(defn start-system []
  (let [system (component/system-map
                 :psil-engine (comps/map->PsilEngine {}))]
    (component/start-system system)))
(def system (start-system))
(defn get-lookup-table []
  (:variable-lookup-map (:psil-engine system)))

(defn run [file]
  (->> file
       io/resource
       slurp
       parse-raw-source
       parse-source
       (eval-program (get-lookup-table))))

(defn run-src [source-str]
  (->> source-str
       parse-raw-source
       parse-source
       (eval-program (get-lookup-table))))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (run (first args)))


(comment
  (->> "123"
       run-src)
  ;=> 123

  (->>
    "(+)"
    run-src)
  ;=> 0

  (->>
    "(*)"
    run-src)
  ;=> 1

  (->>
    "(+ 1)"
    run-src)
  ;=> 1

  (->>
    "(* 1)"
    run-src)
  ;=> 1

  (->>
    "(* 1 1 1 1 2)"
    run-src)
  ;=> 2
  (->>
    "(* 2 (+ 1 2 3) 3)"
    run-src)
  ;=> 36

  (->>
    "(* 2 (+ 1 ) 3)"
    run-src)
  ;=> 6

  (->>
    "(* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))"
    run-src)
  ;=> 192

  (->>
    "(* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))"
    run-src)
  ;=> 192

  (->>
    "(bind a (* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2))))))"
    run-src)
  ;=> 192

  (->>
    "(bind a (* 2 (+ 1 2) (* 2 (* 2 (* 2 (* 2 2)))))) (+ 8 a)"
    run-src)
  ;=> 200

  (->>
    "(bind a 1) (bind b 2) (+ 7 a b)"
    run-src)
  ;=> 10

  (->>
    "(+ 1\n    (* 2 3)\n    (* 4 2))"
    run-src)
  ;=> 15

  (->>
    "(bind radius 12)"
    run-src)
  ;=> 12

  (->>
    "(bind length 10)\n(bind breadth 10)\n(* length breadth)"
    run-src)
  ;=> 100

  (->>
    "(bind length 10) (+ 1 2 3 4) (bind breadth 10)\n(* length breadth)"
    run-src)
  ;=> 100

  (->>
    "(bind length 10)\n(bind breadth (+ length 1))\n(bind length 11)\n(* length breadth)"
    run-src)
  ;=> 121

  (->>
    "(bind a 10)\n(bind b a)\n(bind a 11)\n(+ a b)"
    run-src)
  ;=> 21

  (->>
    "(bind a 1) (bind b 2) (+ c a b)"
    run-src)
  ;=> Invalid Program

  (->>
    "(bind a a) (bind b 2) (+ c a b)"
    run-src)
  ;=> Invalid Program

  (->>
    "(bind a a) (bind b 2) (+ 1)"
    run-src)
  ;=> Invalid Program

  (->>
    "()"
    run-src)
  ;=> Invalid Program

  (->>
    "(1)"
    run-src)
  ;=> Invalid Program

  (->>
    "a"
    run-src)
  ;=> Invalid Program

  (->>
    "( a +"
    run-src)
  ;=> Invalid Program

  (->>
    "( 1 + )"
    run-src)
  ;=> Invalid Program

  (->>
    "(1 2 3 4 5)"
    run-src)
  ;=> Invalid Program

  (->>
    "(+ 1) 2)"
    run-src)
  ;=> Invalid Program

  (->>
    "(+ (* 1 2) + (* 3 4))"
    run-src)
  ;=> Invalid Program



  )